datafusion/expr-common/src/type_coercion/binary.rs

```diff
@@ -777,29 +777,20 @@ pub fn binary_numeric_coercion(
         (_, Float32) | (Float32, _) => Some(Float32),
         // The following match arms encode the following logic: Given the two
         // integral types, we choose the narrowest possible integral type that
-        // accommodates all values of both types. Note that some information
-        // loss is inevitable when we have a signed type and a `UInt64`, in
-        // which case we use `Int64`;i.e. the widest signed integral type.
-
-        // TODO: For i64 and u64, we can use decimal or float64
-        // Postgres has no unsigned type :(
-        // DuckDB v.0.10.0 has double (double precision floating-point number (8 bytes))
-        // for largest signed (signed sixteen-byte integer) and unsigned integer (unsigned sixteen-byte integer)
+        // accommodates all values of both types. Note that to avoid information
+        // loss when combining UInt64 with signed integers we use Decimal128(20, 0).
+        (Decimal128(20, 0), _)
+        | (_, Decimal128(20, 0))
```

Comment:
This looks not correct. For example, combining `Decimal128(20, 0)` with `Decimal128(30, 0)`  should not result in `Decimal128(20, 0)`

Reply:
I think when both types are decimal they are handled above before this match, when calling the `decimal_coercion` function.

Reply:
I rechecked it, and `decimal_coercion` already covers them in https://github.com/apache/datafusion/blob/2f283277d523404401b206e5b98819a0126341b8/datafusion/expr-common/src/type_coercion/binary.rs#L932-L940
Although it doesn't handle unsigned integer types, we can supplement it there, maybe as a follow-up PR.

Reply:
So I think we shouldn't combine decimal with integer types here because `decimal_coercion` has already handled it.

Reply:
I tried that initially but then it would not handle `UInt64` and `Decimal128(20, 0)`: 

```
Cannot infer common argument type for comparison operation UInt64 = Decimal128(20, 0)
```

So maybe it would be best to add new arms to the `coerce_numeric_type_to_decimal` to include unsigned integers as well?

```rust
    match numeric_type {
        Int8 => Some(Decimal128(3, 0)),
        Int16 => Some(Decimal128(5, 0)),
        Int32 => Some(Decimal128(10, 0)),
        Int64 => Some(Decimal128(20, 0)),
        Float32 => Some(Decimal128(14, 7)),
        Float64 => Some(Decimal128(30, 15)),
        _ => None,
    }
```

Reply:
> So maybe it would be best to add new arms to the `coerce_numeric_type_to_decimal` to include unsigned integers as well?

I think so.

Reply:
Done.